wsagent/che-core-api-git/src/main/java/org/eclipse/che/api/git/CachedGitStatusProvider.java

```diff
@@ -0,0 +1,290 @@
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.FileTime;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.inject.Inject;
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.model.workspace.config.ProjectConfig;
+import org.eclipse.che.api.core.notification.EventService;
+import org.eclipse.che.api.fs.server.PathTransformer;
+import org.eclipse.che.api.git.exception.GitException;
+import org.eclipse.che.api.git.shared.FileChangedEventDto;
+import org.eclipse.che.api.git.shared.Status;
+import org.eclipse.che.api.git.shared.StatusChangedEventDto;
+import org.eclipse.che.api.project.server.ProjectManager;
+import org.eclipse.che.api.project.server.VcsStatusProvider;
+import org.eclipse.che.api.project.server.impl.RootDirPathProvider;
+import org.eclipse.che.api.project.server.notification.ProjectCreatedEvent;
+import org.eclipse.che.api.project.server.notification.ProjectDeletedEvent;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Git implementation of {@link VcsStatusProvider} based on a {@link Map} which contains statuses of
+ * all workspace projects. The map is updated by Git events and by checking the file's modification
+ * time to update the map if the files were changed bypassing the file-watchers e.g. the file wasn't
+ * opened neither in the project explorer tree, neither in the editor, but was changed in the
+ * terminal.
+ *
+ * @author Igor Vinokur
+ */
+public class CachedGitStatusProvider implements VcsStatusProvider {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CachedGitStatusProvider.class);
+
+  private final GitConnectionFactory gitConnectionFactory;
+  private final PathTransformer pathTransformer;
+  private final ProjectManager projectManager;
+  private final RootDirPathProvider rootDirPathProvider;
+  private final EventService eventService;
+  private final Map<String, Status> statusCache;
+  private final Map<String, FileTime> projectFiles;
+
+  @Inject
+  public CachedGitStatusProvider(
+      GitConnectionFactory gitConnectionFactory,
+      PathTransformer pathTransformer,
+      ProjectManager projectManager,
+      RootDirPathProvider rootDirPathProvider,
+      EventService eventService) {
+    this.gitConnectionFactory = gitConnectionFactory;
+    this.pathTransformer = pathTransformer;
+    this.projectManager = projectManager;
+    this.rootDirPathProvider = rootDirPathProvider;
+    this.eventService = eventService;
+    this.statusCache = new HashMap<>();
+    this.projectFiles = new HashMap<>();
+
+    subscribeToEvents();
+    collectProjectFiles(rootDirPathProvider.get());
+  }
+
+  private void collectProjectFiles(String root) {
```

Comment:
I propose add @PostConstruct and remove call this method from constructor

Reply:
Done